File: TAC.java
import java.util.Scanner;

public class TAC {
    public static void main(String[] args) {
        // Prompt the user for an input expression
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter an arithmetic expression: ");
        String expr = scanner.nextLine();
        
        // Generate and print the corresponding three-address code
        generateTAC(expr);
    }

    public static void generateTAC(String expr) {
        // Split the input expression into individual tokens
        String[] tokens=expr.split("\\s+");
        int tCount=1;
        String tPrefix="t";
        String code="";
        for (int i = 0; i < tokens.length; i++) {
            if(i==0){
                code+=tPrefix+tCount+" = " +tokens[i]+"\n"; //current
            }else if(i==1){
                tCount++;
                    //current                              previous                                  current
                code+=tPrefix+tCount+" = " +tokens[i]+"\n"+tPrefix+(tCount-1)+"  " +tokens[i-1]+" "+tPrefix+tCount+" " +tokens[i]+"\n";
            }else{
                tCount++; //previous                         current                                  previous
                code+=tPrefix+(tCount-1)+"  " +tokens[i-1]+" "+tPrefix+tCount+"\n" +tPrefix+tCount+" = "+ tPrefix+(tCount-1) +"\n" ;
            }
        }
    }
}

File: TAC.py
def main():
    input_str = input("Enter an expression: ")

    # Split the expression into tokens
    tokens = input_str.split()

    # Initialize the TAC instruction list
    tac_instructions = []

    # Initialize the temporary variable count
    temp_count = 1

    # Initialize the index for the current token
    i = 0

    # Loop through the tokens to generate TAC instructions
    while i < len(tokens):
        token = tokens[i]

        # If the token is an operator ('+', '-', '*', '/', '%')
        if token in ['+', '-', '*', '/', '%']:
            # Get the operands
            operand1 = tokens[i - 1]
            operand2 = tokens[i + 1]

            # Generate the TAC instruction and append it to the list
            tac_instructions.append(f"t{temp_count} = {operand1} {token} {operand2}")

            # Replace the operands with the temporary variable in the expression
            tokens[i - 1] = f"t{temp_count}"
            del tokens[i:i+2]

            # Increment the temporary variable count
            temp_count += 1

            # Adjust the index to skip the processed operands
            i -= 1

        # Increment the index to process the next token
        i += 1

    # Generate the final assignment instruction
    final_result = "result"
    if len(tokens) > 1:
        # If there are remaining tokens, concatenate them with the last temporary variable
        final_result = f"t{temp_count} = {' '.join(tokens)}"
        temp_count += 1

    # Print the TAC instructions
    print("Three Address Code (TAC):")
    for instruction in tac_instructions:
        print(instruction)
    print(final_result)


if __name__ == "__main__":
    main()